extensions/resteasy-reactive/rest/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/security/EagerSecurityContext.java

```diff
@@ -148,7 +148,9 @@ public SecurityIdentity apply(SecurityCheckWithIdentity checkWithIdentity) {
                             if (eventHelper.fireEventOnSuccess()) {
                                 eventHelper.fireSuccessEvent(new AuthorizationSuccessEvent(newIdentity,
                                         AbstractPathMatchingHttpSecurityPolicy.class.getName(),
-                                        Map.of(RoutingContext.class.getName(), routingContext)));
+                                        Map.of(RoutingContext.class.getName(), routingContext,
+                                                AuthorizationSuccessEvent.SECURITY_IDENTITY_AUGMENTED,
```

Comment:
Will this event be fired only once ?

Reply:
Once by JAX-RS Http perms. that is if you have 5 roles policies matched, it will be fired only once.

Reply:
If someone however augment identity inside non-JAXRS HTTP Perms, then there of course is different event. Because it is different authorization. This is authorization.

Reply:
So, there will be AuthenticationSuccess event and then there could be a 2nd AuthenticationSuccess event with the boolean flag the identity has been augmented ?

Reply:
So there will **always** be only one `AuthenticationSuccessEvent`, I didn't changed that and won't.

Reply:
@michalvavrik I'm sorry if I'm still confused. So we will have a single AuthenticationSuccess event only as before., this is clear. But how many AuthorizationSuccess events can be fired now ? If it is more than 1 then I'm not sure it is necessary, which I what I said offline, if we have 3 places where the identity is augmented then it will lead to 3 events without providing no actual augmentation difference, in case of Otel, you'll have the span updated 3 times as part of the single request...

Reply:
> But how many AuthorizationSuccess events can be fired now ? 

Theoretically, if and only if user configured all these authorization measures, in following cases:

- once after all `HTTPSecurityPolicy`s inside Vert.x HTTP extension are finished
- once after all `HTTPSecurityPolicy`s inside Quarkus REST and Quarkus RESTEasy Classic extension are finished
- once after all `SecurityCheck`s (standard security annotations or default JAX-RS security) are finished

I want to stress that there is no HTTP request **end** hook we could use to unify Vert.x with Jax-RS events. To my knowledge that is not technically possible. So many things could go wrong if we don't fire the event after authorization is finished. Also right now, you can see the tests inside gRPC which shows that it needs to work for other stacks as well. And it does.

> If it is more than 1 then I'm not sure it is necessary which I what I said offline, if we have 3 places where the identity is augmented then it will lead to 3 events without providing no actual augmentation difference,

Only `HTTPSecurityPolicy`s can augment identity. So it is in theory 2 places. However, in practice it must be one @sberyozkin , just think about this:

_JAX-RS specific HTTPSecurityPolicy_ exists so that we can postpone authentication. Therefore the idea that user will setup first HTTPSecurityPolicy for HTTP request and then select OIDC tenant via `@Tenat` annotation is illegal and we document this. Therefore there in practice there is one authorization event fired that contains the augmented event property.

I am sorry, it is complicated matter. Feel free to ask again.

> in case of Otel, you'll have the span updated 3 times as part of the single request...

I will have update span when authentication finishes and if and only if HTTP Security policy augmented the identity. I am no expert on OpenTelemetry, they might reject it in the end, but I read documentation https://opentelemetry.io/docs/specs/semconv/general/attributes/#general-identity-attributes and it seems like `best-effort` feature. My plan will provide with near to 100 % reliability in case of proactive authentication and well, when proactive authentication is disabled when authN/authZ happens. It could certainly get rejected, that's life :-)

P.S. I plan to document custom JAX-RS SecurityContext as not supported with such OTel feature because we didn't agree on how to solve this and discussion in this PR is big already.

Reply:
@michalvavrik I'm sorry for a delay, I'm just struggling a bit to make sure I understand everything. I'll get back shortly.

>P.S. I plan to document custom JAX-RS SecurityContext as not supported with such OTel feature because we didn't agree on how to solve this and discussion in this PR is big already.

I have to admit I barely recall now what we discussed it around that :-)

Reply:
@michalvavrik So, am I correct in assuming the following will be fired with this PR during the successful security request when `@RolesAllowed` is used:

* `AuthenticationSuccessEvent` is fired once - when SecurityIdentity is created, even if it includes one or more `SecurityIdentityAugmentor`
* `AuthorizationSuccessEvent`  is fired once - when the `@RolesAllowed` check is passed

Now, lets assume that instead of `@RolesAllowed`, we add a single HTTP Policy. What happens now, I'll start:

* `AuthenticationSuccessEvent` is fired once - when SecurityIdentity is created, even if it includes one or more `SecurityIdentityAugmentor`
* And what else is fired and how many times ?

Thanks

Reply:
> @michalvavrik So, am I correct in assuming the following will be fired with this PR during the successful security request when `@RolesAllowed` is used:
> 
> * `AuthenticationSuccessEvent` is fired once - when SecurityIdentity is created, even if it includes one or more `SecurityIdentityAugmentor`
> * `AuthorizationSuccessEvent`  is fired once - when the `@RolesAllowed` check is passed
> 

Exactly, we have tons of tests for this so this behavior is assured.

> Now, lets assume that instead of `@RolesAllowed`, we add a single HTTP Policy with the roles allowed settings. What happens now, I'll start:
> 
> * `AuthenticationSuccessEvent` is fired once - when SecurityIdentity is created, even if it includes one or more `SecurityIdentityAugmentor`

Yes.

> * And what else is fired and how many times ?
> 

HTTP Security Policy (whatever kind ALL / JAXRS) will be fired. If of interest, I can lookup some test coverage we have (on demand).

> Thanks

Same here.

> I have to admit I barely recall now what we discussed around that :-)

I had to re-read whole PR because I was a short on details :-) 

Let me stress one thing @sberyozkin , this PR doesn't change when the events are fired or how many times (though your questions are valid and you can totally continue in them!). It changes 2 things:

1. when HTTP Security Policy denied access, we didn't trigger the event with the identity (possibly) produced by the HTTP Security Policies
2. moves `quarkus.http.auth.roles-mapping` to the authentication phase because this way, the augmented identity will always be present in the authentication event. that is needed because there is no authorization event for teh `quarkus.http.auth.roles-mapping` because it is not an authorization feature. That is where the augmentors are as well, so I don't see why not?

Reply:
> @michalvavrik So, we only have HTTP security checks producing the AuthorizationSuccessEvent or failure events with SECURITY_IDENTITY_AUGMENTED status

yep

> , but how do you know HTTP security has augmented the identity in case of the check denying or allowing access ?

Identity has been augmented if:

1.  authentication has happened; e.g. if only deferred identity had been available when proactive auth is disabled, we have it now available on the RoutingContext because that is where we put it:

- https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java#L300

- https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java#L314

- https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java#L356

If authentication had failed, there is no point to try to figure whether it is augmented, so we only care about "auth success" scenarios where the identity is available

2. the identity is not the very same instance (no equals method, `=`)

I was rewriting these parts lately because of security events, but I didn't change IF condition, it is in main for long before me. Please have a look:

https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/AbstractHttpAuthorizer.java#L103

So in this PR, I wanted to check even for "authorization failure" scenarios.